### PR TITLE
polish(rjy): polish notes for q-learning

### PR DESCRIPTION
**Pull request by nighood**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/model/template/q_learning.py [hunk 1]
```diff
@@ -663,6 +663,9 @@ def reshape(d):
                 d = d.reshape(T, B, *d.shape[1:])
             return d
 
+        # NOTE(rjy): the initial input shape will be (T,B,N),
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:666]: `(T, B, N)`

File: ding/model/template/q_learning.py [hunk 2]
```diff
@@ -663,6 +663,9 @@ def reshape(d):
...
+        #            means encoder or head should process T*B trajectories,
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:667]: process B trajectorys, each trajectory has T timestep

File: ding/model/template/q_learning.py [hunk 3]
```diff
@@ -663,6 +663,9 @@ def reshape(d):
...
+        #            so here we need such fn to reshape for encoder or head
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:668]: but T and B dimension can be both treated as `batch_size` in encoder and head, i.e., independent and parallel processing, se we reshape...

File: ding/model/template/q_learning.py [hunk 4]
```diff
@@ -792,6 +795,7 @@ def forward(self, inputs: Dict, inference: bool = False, saved_state_timesteps:
         x, prev_state = inputs['obs'], inputs['prev_state']
         # for both inference and other cases, the network structure is encoder -> rnn network -> head
         # the difference is inference take the data with seq_len=1 (or T = 1)
+        # NOTE(rjy): ? in most situations, set inference=True when evaluate and inference=False when training
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:798]: why `?`

File: ding/model/template/q_learning.py [hunk 5]
```diff
@@ -806,16 +810,25 @@ def forward(self, inputs: Dict, inference: bool = False, saved_state_timesteps:
             x['next_state'] = next_state
             return x
         else:
+            # NOTE(rjy): in r2d2, 
+            # 1) data['burnin_nstep_obs'] = data['obs'][:bs + self._nstep]
+            # 2) data['main_obs'] = data['obs'][bs:-self._nstep]
+            # 3) data['target_obs'] = data['obs'][bs + self._nstep:]
+            # NOTE(rjy): why len must be [3,5] ?
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:817]: `(T, B, N)` or `(T, B, C, H, W)`

File: ding/model/template/q_learning.py [hunk 6]
```diff
@@ -806,16 +810,25 @@ def forward(self, inputs: Dict, inference: bool = False, saved_state_timesteps:
             x['next_state'] = next_state
             return x
         else:
+            # NOTE(rjy): in r2d2,
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:813]: why add R2D2 information here

File: ding/model/template/q_learning.py [hunk 7]
```diff
@@ -806,16 +810,25 @@ def forward(self, inputs: Dict, inference: bool = False, saved_state_timesteps:
...
             if self.res_link:
                 a = x
+            # NOTE(rjy) lstm_embedding stores all hidden_state
             lstm_embedding = []
             # TODO(nyz) how to deal with hidden_size key-value
             hidden_state_list = []
             if saved_state_timesteps is not None:
                 saved_state = []
             for t in range(x.shape[0]):  # T timesteps
+                # NOTE(rjy) why use x[t:t + 1] but not x[t]?
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:831]: `x[t:t+1]` can keep original dimension, like this
```python
a = torch.randn(4, 3)
assert a[0:1].shape == (1, 3)
assert a[0].shape == (3, )
```

File: ding/model/template/q_learning.py [hunk 8]
```diff
@@ -827,14 +840,19 @@ def forward(self, inputs: Dict, inference: bool = False, saved_state_timesteps:
...
+            # NOTE(rjy): x['next_state'] is the hidden state of the last timestep inputted to lstm  
             # the last timestep state including the hidden state (h) and the cell state (c)
             # shape: {list: B{dict: 2{Tensor:(1, 1, head_hidden_size}}}
             x['next_state'] = prev_state
             # all hidden state h, this returns a tensor of the dim: seq_len*batch_size*head_hidden_size
             # This key is used in qtran, the algorithm requires to retain all h_{t} during training
-            x['hidden_state'] = torch.cat(hidden_state_list, dim=-3)
+            # NOTE(rjy) why dim=-3? for the hidden_states' shape are [1,batch_size,head_hidden_size], so dim=-3.
+            #           but at begining, we assert 'len(x.shape) in [3, 5], x.shape'? how about x.shape=5?
+            x['hidden_state'] = torch.cat(hidden_state_list, dim=-3)
```
Comment (PaParaZz1) [inline ding/model/template/q_learning.py:851]: dim=0

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/model/template/q_learning.py (hunk 6 above)
Comment (nighood) [reply ding/model/template/q_learning.py:813]: without r2d2, I can't understand why we need saved_state here and which state should be saved

Comment (PaParaZz1) [reply ding/model/template/q_learning.py:815]: add this reason in comments here

**Review by PaParaZz1 — APPROVED** [review]
